File: 0065-valid-number/0065-valid-number.py
class Solution:
    def isNumber(self, s: str) -> bool:
        digit = False
        dot = False
        exp = False
        
        for i, ch in enumerate(s):
            
            if ch.isdigit():
                digit = True
            
            elif ch in ['+', '-']:
                
                if i > 0 and s[i-1] not in ['e', 'E']:
                    return False
            
            elif ch == '.':
                
                if dot or exp:
                    return False
                dot = True
            
            elif ch in ['e', 'E']:
                
                if exp or not digit:
                    return False
                exp = True
                digit = False   
            else:
                return False
        
        return digit

File: 0065-valid-number/0065-valid-number.cpp
// LeetCode 65 - Valid Number
// https://leetcode.com/problems/valid-number/
// Author: muteeba4266

#include <string>
using namespace std;

class Solution {
public:
    bool isNumber(string s) {
        bool digit = false;
        bool dot = false;
        bool exp = false;

        for (int i = 0; i < s.length(); i++) {
            
            if (isdigit(s[i])) {
                digit = true;
            }
            
            else if (s[i] == '+' || s[i] == '-') {
                if (i > 0 && s[i-1] != 'e' && s[i-1] != 'E')
                    return false;
            }
            
            else if (s[i] == '.') {
                if (dot || exp)
                    return false;
                dot = true;
            }
            
            else if (s[i] == 'e' || s[i] == 'E') {
                if (exp || !digit)
                    return false;
                exp = true;
                digit = false;
            }
            
            else {
                return false;
            }
        }

        return digit;
    }
};